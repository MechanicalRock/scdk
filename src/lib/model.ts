// types for rendering the constructs
export type Product = {
    Name: string;
    ProvisioningArtifacts: ProvisioningArtifact[];
};

export type ProvisioningArtifact = {
    Name: string;
    Parameters: ProvisioningArtifactParameter[];
    Outputs: {
        OutputName: string;
        OutputDescription?: string;
    }[];
};

export type ProvisioningArtifactParameter = {
    ParameterName: string;
    ParameterType: string;
    ParameterRequired: boolean;
    ParameterDescription?: string;
};

export type Portfolio = {
    Products: Product[];
};

// Types for parsing CloudFormation templates
export type Template = {
    Parameters: { [key: string]: Parameter };
    Outputs: { [key: string]: Output };
};

export type Parameter = {
    Type: ParameterType;
    Description?: string;
    Default?: string;
};

export type ParameterType = string;

export type Output = {
    Description?: string;
};
