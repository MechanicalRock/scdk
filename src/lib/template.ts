import { Environment, FileSystemLoader } from "nunjucks";
import pascalcase from "pascalcase";
import { join } from "path";
import { promises as fs, existsSync, mkdirSync } from "fs";
import { Product, Portfolio } from "./model";
import { major } from "semver";

const njs = new Environment(new FileSystemLoader(join(__dirname, "../..", "templates")));

export async function writeTemplates(portfolio: Portfolio, baseDir: string): Promise<void[]> {
    if (!existsSync(baseDir)) {
        mkdirSync(baseDir);
    }
    return await Promise.all(
        Object.entries(getTemplates(portfolio)).map(([filename, template]) =>
            fs.writeFile(join(baseDir, filename), template),
        ),
    );
}

export function getTemplates(portfolio: Portfolio): { [key: string]: string } {
    const index = {
        ["index.ts"]: renderPortfolio(portfolio),
    };
    const fileContents = portfolio.Products.map((p) => ({
        [`${pascalcase(p.Name)}.ts`]: renderProduct(p),
    }));
    return Object.assign(index, ...fileContents);
}

export const renderProduct = (product: Product): string => {
    return njs.render("product.ts.njk", { ...product, pascalcase, major });
};

export const renderPortfolio = (portfolio: Portfolio): string => {
    return njs.render("portfolio.ts.njk", { ...portfolio, pascalcase });
};
