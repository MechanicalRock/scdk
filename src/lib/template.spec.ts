import { renderPortfolio, renderProduct } from "./template";
import { Portfolio } from "./model";
import chai from "chai";
import chaiJestSnapshot from "chai-jest-snapshot";
chai.use(chaiJestSnapshot);

const portfolio: Portfolio = {
    Products: [
        {
            Name: "cloud-pipeline",
            ProvisioningArtifacts: [
                {
                    Name: "v1",
                    Parameters: [
                        {
                            ParameterName: "input_one",
                            ParameterType: "String",
                            ParameterRequired: true,
                        },
                        {
                            ParameterName: "input_two",
                            ParameterType: "String",
                            ParameterRequired: false,
                        },
                    ],
                    Outputs: [
                        {
                            OutputName: "output_one",
                        },
                    ],
                },
            ],
        },
        {
            Name: "cloud-website",
            ProvisioningArtifacts: [
                {
                    Name: "v1",
                    Parameters: [
                        {
                            ParameterName: "input_one",
                            ParameterType: "String",
                            ParameterRequired: false,
                        },
                    ],
                    Outputs: [
                        {
                            OutputName: "output_one",
                        },
                        {
                            OutputName: "output_two",
                        },
                    ],
                },
            ],
        },
    ],
};

describe("Given a portfolio definition", function () {
    before(function () {
        chaiJestSnapshot.resetSnapshotRegistry();
    });

    beforeEach(function () {
        chaiJestSnapshot.configureUsingMochaContext(this);
    });

    describe("When I render the template", function () {
        const template = renderPortfolio(portfolio);

        // it("Then the rendered file should be valid typescript", function () {});

        it("Then the rendered file should match the snapshot", function () {
            chai.expect(template).to.matchSnapshot();
        });
    });
});

describe("Given a product definition", function () {
    describe("When I render the template", function () {
        const template = renderProduct(portfolio.Products[0]);

        // it("Then the rendered file should be valid typescript", function () {});

        it("Then the rendered file should match the snapshot", function () {
            chai.expect(template).to.matchSnapshot();
        });
    });
});
