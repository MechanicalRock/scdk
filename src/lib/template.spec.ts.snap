// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a portfolio definition When I render the template Then the rendered file should match the snapshot 1`] = `
"export * as CloudPipeline from \\"./CloudPipeline\\";
export * as CloudWebsite from \\"./CloudWebsite\\";
"
`;

exports[`Given a portfolio definition When I render the template Then the rendered file should match the snapshot 2`] = `
"import { Construct, Reference } from \\"@aws-cdk/core\\";
import { CfnCloudFormationProvisionedProduct } from \\"@aws-cdk/aws-servicecatalog\\";
interface v1Props {
    input_one: string;
    input_two?: string;
}

export class V1 extends Construct {
    readonly output_one: Reference;

    constructor(scope: Construct, id: string, props: v1Props) {
        super(scope, id);

        const provisioningParameters = Object.entries(props).map(([key, value]) => ({ key, value }));

        const resource = new CfnCloudFormationProvisionedProduct(this, \\"Resource\\", {
            productName: \\"cloud-pipeline\\",
            provisioningArtifactName: \\"v1\\",
            provisioningParameters,
        });
        this.output_one = resource.getAtt(\\"outputs.output_one\\");
    }
}
"
`;
